Render transparent objects rear faces before front faces

thehans opened the pull request:
for issue #1496 

TODO:
- [x] Consider performance impact

thehans commented:
Its currently a little difficult to directly measure rendering performance.  I'm thinking it might be good to have some performance test functionality built in, where enabling it would render frames as quickly as possible, and spit out an FPS count.  

This might make sense to put the FPS readout as an added feature to the animation settings, since it doesn't necessarily always hit the FPS that you set it to.  
And then maybe a checkbox next to FPS input box for "Max" which would completely disable any delay between frames.

Also needed is some particularly gnarly test case that gives the renderer something to chew on.

Thoughts?

kintel commented:
See #1561 - to support higlighting unexpected backfaces, we render _front_ faces first, then backfaces.
I think the #1561 workflow is more important, but I wanted to raise that conflict here, in case someone else have other arguments.

thehans commented:
@kintel I rebased this on current master and fixed conflicts.  In doing this I cleaned up the behavior of my code so that opaque object rendering should be unchanged.  I think the changes needed for #1561 can go within that conditional branch for handling opaque objects without any problem.

I also think it is a harmless performance enhancement to cull the front faces of negative shapes, so I did that, which, should hopefully make up for any performance impact of doing two passes on transparent shapes.

kintel commented on src/OpenCSGRenderer.cc:
```diff
@@ -123,10 +123,22 @@ void OpenCSGRenderer::renderCSGProducts(const CSGProducts &products, GLint *shad
 			glMultMatrixd(csgobj.leaf->matrix.data());
-			render_surface(csgobj.leaf->geom, csgmode, csgobj.leaf->matrix, shaderinfo);
+			
+			if (setColor(colormode, c.data(), shaderinfo)) {
```
I think it would be cleaner to move the alpha check out of `setColor()` and do it manually here.

thehans replied:
The problem is that setColor doesn't unconditionally set the color that you give it from this scope. It depends if the object is highlighted, etc.  Putting the check here would mean basically duplicating all the code of setColor.   This function previously had void return type, I figured a boolean return was the simplest way to get that info out of the function.  Would it be preferrable to return the full Color4f basecol instead?  Then check the alpha of that here?

kintel replied:
Yes, that's a cleaner solution.
Better would be to e.g. split out those calculations from `setColor()`, but that's probably out of scope for this PR.

kintel commented:
Humm, somehow these tests fail on the Mac CI:
```
The following tests FAILED:
	322 - cgalpngtest_polyhedron-nonplanar-tests (Failed)
	643 - csgpngtest_polyhedron-nonplanar-tests (Failed)
	895 - monotonepngtest_polyhedron-nonplanar-tests (Failed)
	918 - cgalstlcgalpngtest_polyhedron-nonplanar-tests (Failed)
```

Any ideas?
I'll have to open the PR in an OpenSCAD-owned branch to get the test result upload from the CI to work..

kintel commented:
@thehans It looks like the test failures are unrelated to this PR: Master also fails now. I suspect homebrew update on TravisCI from 4.9 to 4.11 to be the issue -> we should fix that (CGAL 4.11 support) separately.
